src/main/java/net/neoforged/neoforge/common/util/InsertableLinkedOpenCustomHashSet.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.util;
+
+import it.unimi.dsi.fastutil.Hash;
+import it.unimi.dsi.fastutil.HashCommon;
+import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenCustomHashSet;
+import net.neoforged.neoforge.common.util.strategy.BasicStrategy;
+import net.neoforged.neoforge.common.util.strategy.IdentityStrategy;
+
+/**
+ * Special linked hash set that allow changing the order of its entries and is strict to throw if attempting to add an entry that already exists.
+ * Requires a strategy for the hashing behavior. Use {@link BasicStrategy#BASIC} or {@link IdentityStrategy#IDENTITY} if no special hashing needed.
+ */
+public class InsertableLinkedOpenCustomHashSet<T> extends ObjectLinkedOpenCustomHashSet<T> {
+    private static final long LINK_BITS = 0xFFFFFFFFL;
+    private static final long LINK_BIT_SPACE = 32L;
+    private static final long NEXT_LINK = LINK_BITS;
+    private static final long PREV_LINK = LINK_BITS << LINK_BIT_SPACE;
+
+    public InsertableLinkedOpenCustomHashSet() {
+        super(BasicStrategy.BASIC);
+    }
+
+    public InsertableLinkedOpenCustomHashSet(Hash.Strategy<? super T> strategy) {
+        super(strategy);
+    }
+
+    public boolean addAfter(T insertAfter, T element) {
```

Comment:
Some comments on what this method is doing would be helpful. Same for `addBefore`.